core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -675,35 +827,115 @@ private Optional<Throwable> closeAndAddSuppressed(Optional<Throwable> existingFa
 
         private boolean processEvents()
         {
+            boolean eventDebugInfoRecorded = false;
             try {
-                Event event = eventQueue.poll(EVENT_PROCESSING_ENFORCED_FREQUENCY_MILLIS, MILLISECONDS);
-                if (event == null) {
-                    return true;
+                try {
+                    Event event = eventQueue.poll(EVENT_PROCESSING_ENFORCED_FREQUENCY_MILLIS, MILLISECONDS);
+                    if (event == null) {
+                        return true;
+                    }
+                    eventBuffer.add(event);
                 }
-                eventBuffer.add(event);
-            }
-            catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RuntimeException(e);
-            }
-
-            while (true) {
-                // poll multiple events from the queue in one shot to improve efficiency
-                eventQueue.drainTo(eventBuffer, EVENT_BUFFER_CAPACITY - eventBuffer.size());
-                if (eventBuffer.isEmpty()) {
-                    return true;
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
                 }
-                for (Event e : eventBuffer) {
-                    if (e == Event.ABORT) {
-                        return false;
+
+                while (true) {
+                    // poll multiple events from the queue in one shot to improve efficiency
+                    eventQueue.drainTo(eventBuffer, EVENT_BUFFER_CAPACITY - eventBuffer.size());
+                    if (eventBuffer.isEmpty()) {
+                        return true;
                     }
-                    if (e == Event.WAKE_UP) {
-                        continue;
+
+                    for (Event e : eventBuffer) {
+                        if (e == Event.ABORT) {
+                            return false;
+                        }
+                        if (e == Event.WAKE_UP) {
+                            continue;
+                        }
+                        e.accept(this);
                     }
-                    e.accept(this);
+
+                    eventDebugInfoRecorded |= recordEventsDebugInfo(eventBuffer);
```

Comment:
move inside `for (Event e : eventBuffer) {` loop;
make `recordEventsDebugInfo` accept single event